### PR TITLE
CLOUDP-101056 Percentage time spent in proxy metrics

### PR DESCRIPTION
**Pull request by pstefek-mongo**
Jira: https://jira.mongodb.org/browse/CLOUDP-101056
EVG: https://spruce.mongodb.com/version/614a397d1e2d177a88c22067/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

### CODE REVIEW

File: proxy_session.go [hunk 1]
```diff
@@ -542,6 +546,10 @@ func (ps *ProxySession) doLoop(mongoConn *MongoConnectionWrapper, retryError *Pr
 			if hookErr != nil {
 				ps.proxy.logger.Logf(slogger.WARN, "failed to observe dbRoundTripHook %v", hookErr)
 			}
+			hookErr = percentageTimeSpentInProxy.ObserveWithLabels(totalProxyProcessingTimeElaspsed/(totalDBRoundtripTime+totalProxyProcessingTimeElaspsed), map[string]string{"operation": metricOperationType})
```
Comment (pstefek-mongo) [inline proxy_session.go:549]: Can you just give me a sanity check on this calculation?

Comment (TomerYakir) [reply proxy_session.go:549]: lgtm - just to be super-safe, let's add a validation that totalDBRoundtripTime+totalProxyProcessingTimeElaspsed is > 0

Comment (TomerYakir) [reply proxy_session.go:549]: btw note that we need to multiply by 100 if we want to represent this as percentage

**Comment by pstefek-mongo** [comment]
EVG: https://evergreen.mongodb.com/version/614b391f3e8e862970eb949a?redirect_spruce_users=true

File: proxy_session.go (hunk 1 above)
Comment (pstefek-mongo) [reply proxy_session.go]: I'm okay keeping it a float

File: proxy_session.go [hunk 2]
```diff
@@ -542,6 +546,15 @@ func (ps *ProxySession) doLoop(mongoConn *MongoConnectionWrapper, retryError *Pr
 			if hookErr != nil {
 				ps.proxy.logger.Logf(slogger.WARN, "failed to observe dbRoundTripHook %v", hookErr)
 			}
+			totalTime := (totalDBRoundtripTime + totalProxyProcessingTimeElaspsed)
+			if totalTime < 0 {
```
Comment (TomerYakir) [inline proxy_session.go:550]: my concern was around division by zero.
Let's change to `if totalTime <= 0 {` to capture non-positive cases

Comment (pstefek-mongo) [reply proxy_session.go]: ah okay

File: proxy_session.go [hunk 3]
```diff
@@ -542,6 +546,15 @@ func (ps *ProxySession) doLoop(mongoConn *MongoConnectionWrapper, retryError *Pr
 			if hookErr != nil {
 				ps.proxy.logger.Logf(slogger.WARN, "failed to observe dbRoundTripHook %v", hookErr)
 			}
+			totalTime := (totalDBRoundtripTime + totalProxyProcessingTimeElaspsed)
+			if totalTime <= 0 {
+				ps.proxy.logger.Logf(slogger.WARN, "total time spent in proxy measure is <= 0 with value %v", totalTime)
```
Comment (TomerYakir) [inline proxy_session.go:551]: should this read `total time spent in proxy and database`?

**Review by TomerYakir — APPROVED** [review]
LGTM with a small question/comment

File: proxy_session.go (hunk 3 above)
Comment (pstefek-mongo) [reply proxy_session.go]: true

**Pull request merged**